drivers/spi/spi_eos_s3.c

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (c) 2020 Antmicro <www.antmicro.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT quicklogic_eos_s3_spi
+
+#define LOG_LEVEL CONFIG_SPI_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(spi_eos_s3);
+#include "spi_context.h"
+
+#include <zephyr/sys/sys_io.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/spi.h>
+#include <eoss3_hal_spi.h>
+#include <eoss3_dev.h>
+#include "spi_eos_s3.h"
+#if defined(CONFIG_PINCTRL)
+#include <zephyr/drivers/pinctrl.h>
+#endif
+
+struct spi_eos_s3_data {
+	struct spi_context ctx;
+	SPI_Mode mode;
+};
+
+struct spi_eos_s3_config {
+	SPI_TypeDef *regs;
+#if defined(CONFIG_PINCTRL)
+	const struct pinctrl_dev_config *pincfg;
+#endif
+};
+
+static int spi_eos_s3_configure(SPI_TypeDef *regs,
+				const struct spi_config *config,
+				SPI_Mode mode)
+{
+	uint32_t ctrlr0;
+
+	if (SPI_WORD_SIZE_GET(config->operation) != SPI_WORD_SIZE) {
+		LOG_ERR("Word size must be %d", SPI_WORD_SIZE);
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_CS_ACTIVE_HIGH) {
+		LOG_ERR("CS active high not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_LOCK_ON) {
+		LOG_ERR("Lock On not supported");
+		return -ENOTSUP;
+	}
+
+	if ((config->operation & SPI_LINES_MASK) != SPI_LINES_SINGLE) {
+		LOG_ERR("Only supports single mode");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_TRANSFER_LSB) {
+		LOG_ERR("LSB first not supported");
+		return -ENOTSUP;
+	}
+
+	if (config->operation & SPI_OP_MODE_SLAVE) {
+		LOG_ERR("Slave mode not supported");
+		return -ENOTSUP;
+	}
+
+	/* Disable SPI controller */
+	regs->SSIENR = SSIENR_SSI_DISABLE;
+
+	/* Disable Slave Select */
+	regs->SER = 0;
+
+	/* Find divisor to setup clock rate */
+	uint32_t clock_div = SPI_CLK / config->frequency;
+
+	/* Divisor must be an even number */
+	if (clock_div % 2) {
+		clock_div++;
+	}
+
+	/* This driver doesn't work correctly with clock rate higher than 2MHz.
+	 * 10MHz / 6 ~= 1.6 MHz
+	 */
+	if (clock_div < 6)
+		return -EINVAL;
```

Comment:
Do we need to always increment `clock_div` when it's not an even divisor? What if `clock_div` was 5, got incremented to 6 instead of decremented to 4, and then rejected because it's not `< 6`?

Reply:
> Do we need to always increment clock_div when it's not an even divisor?

This is what the original driver from Antmicro's fork did. I've checked EOS S3 docs, and indeed it is possible to divide the clock only by multiples of 2 (and 2 is the minimum divisor).

> What if clock_div was 5, got incremented to 6 instead of decremented to 4, and then rejected because it's not < 6?

Driver rejects divisors smaller than 6, not bigger.

Reply:
Fair enough, thanks for elaborating on it